react/components/ModalContext.tsx

```diff
@@ -0,0 +1,86 @@
+import React, { createContext, useReducer, useContext } from 'react'
+
+interface State {
+  open: boolean
+}
+
+const DEFAULT_STATE = {
+  open: false,
+}
+
+function mergeStates(state?: Partial<State>) {
+  return {
+    ...DEFAULT_STATE,
+    ...state,
+  }
+}
+
+interface OpenAction {
+  type: 'OPEN_MODAL'
+}
+
+interface CloseAction {
+  type: 'CLOSE_MODAL'
+}
+
+type Action = OpenAction | CloseAction
+type Dispatch = (action: Action) => void
+
+const ModalStateContext = createContext<State>(DEFAULT_STATE)
+const ModalDispatchContext = createContext<Dispatch | undefined>(undefined)
+
+function modalContextReducer(state: State = DEFAULT_STATE, action: Action) {
+  switch (action.type) {
+    case 'OPEN_MODAL': {
+
+      return {
+        ...state,
+        open: true,
+      }
+    }
+
+    case 'CLOSE_MODAL':
+        return {
+          ...state,
+          open: false,
+        }
+
+    default:
+      return state
+  }
+}
+
+interface ModalContextProviderProps {
+  initialState?: Partial<State>
+}
+
+export const ModalContextProvider: React.FC<ModalContextProviderProps> = props => {
+  const { children, initialState } = props
+  const [state, dispatch] = useReducer(modalContextReducer, mergeStates(initialState))
```

Comment:
Why the usage of `mergeStates`?

Reply:
I was just thinking about the possibility of passing an initial state to the context, but this is probably will never be used, gonna remove it